nhs/data/allocation.py

```diff
@@ -0,0 +1,286 @@
+        └───────────┴──────────┴───────────┴──────────┘
+    """
+    census_type_casted = census.with_columns(
+        pl.col(left_code_col).cast(coords.collect_schema()[right_code_col])
+    )
+
+    return census_type_casted.join(
+        coords, left_on=left_code_col, right_on=right_code_col, how="inner"
+    )
+
+
+@log_entry_exit()
+def sample_census_feature(
+    census: pl.LazyFrame, code_col: str, long_col: str, lat_col: str, feature_col: str
+):
+    """
+    Randomly sample rows from each group in the census LazyFrame using value of `feature_col` as sample size.
+
+    Parameters
+    ----------
+    census: pl.LazyFrame
+        A `LazyFrame` to sample from, with columns `code_col`, `long_col`, `lat_col`,
+        and `feature_col`.
+    code_col : str
+        Column in `census` that contains some identifier/code.
+    long_col : str
+        Column in `census` that contains the longitude.
+    lat_col : str
+        Column in `census` that contains the latitude.
+    feature_col : str
+        Column in `census` for a feature where each row is the
+        sample size for each group.
+
+    Returns
+    -------
+    pl.LazyFrame
+        A `LazyFrame` with the columns `[code_col, long_col, lat_col, feature_col]`
+        where each `code_col` group have N number of randomly sampled rows. N
+        is the value of `feature_col` for each group.
+
+    Examples
+    --------
+    >>> census = pl.LazyFrame({
+    ...     "code_col": ["A", "A", "A", "B"],
+    ...     "long_col": [1.0, 2.0, 3.0, 4.0],
+    ...     "lat_col": [10.0, 20.0, 30.0, 40.0],
+    ...     "feature_col": [4, 4, 4, 2],
+    ... })
+    >>> sample_census_feature(
+    ...     census, "code_col", "long_col", "lat_col", "feature_col"
+    ... ).collect()
+    shape: (6, 4)
+    ┌──────────┬─────────────┬──────────┬─────────┐
+    │ code_col ┆ feature_col ┆ long_col ┆ lat_col │
+    │ ---      ┆ ---         ┆ ---      ┆ ---     │
+    │ str      ┆ bool        ┆ f64      ┆ f64     │
+    ╞══════════╪═════════════╪══════════╪═════════╡
+    │ A        ┆ true        ┆ 1.0      ┆ 10.0    │
+    │ A        ┆ true        ┆ 1.0      ┆ 10.0    │
+    │ A        ┆ true        ┆ 2.0      ┆ 20.0    │
+    │ A        ┆ true        ┆ 3.0      ┆ 30.0    │
+    │ B        ┆ true        ┆ 4.0      ┆ 40.0    │
+    │ B        ┆ true        ┆ 4.0      ┆ 40.0    │
+    └──────────┴─────────────┴──────────┴─────────┘
+    """
+    if census.select(pl.len()).collect().item() == 0:
+        logger.warning("Empty census data provided.")
+        return census.select(pl.col(code_col, long_col, lat_col, feature_col))
+
+    return (
+        census.select(
+            pl.col(code_col, long_col, lat_col, feature_col)
+            # repeat (long, lat) to ensure num coordinates is bigger than sample size
+            .repeat_by(pl.col(feature_col)).flatten()
+        )
+        .filter(
+            pl.int_range(pl.len()).shuffle().over(code_col) < pl.col(feature_col)
```

Comment:
This assigning an index logic will work for the first feature only, the iterative approach needs to track only other columns that are false leftover and then assign the proportion to be true.

So of 100 if 20 are 4 year olds then we have 80 spots left to assign for the 5-10 year olds which only need another 20 and so on...